spec.bs

```diff
@@ -696,26 +698,59 @@ To <dfn>generate and score bids</dfn> given an [=auction config=] |auctionConfig
+1. If |generatedBid|'s [=generated bid/ad component descriptors=] is not null:
+  1. [=set/For each=] |adComponentDescriptor| in |generatedBid|'s
+    [=generated bid/ad component descriptors=]:
+    1. [=set/Append=] |adComponentDescriptor|'s [=ad descriptor/url=] to |adComponentRenderUrls|.
+1. Let |fullSignalsUrl| be the result of [=building trusted scoring signals url=] with |auctionConfig|'s
+  [=auction config/trusted scoring signals url=], |renderUrls|, |adComponentRenderUrls|,
+  |auctionConfig|'s [=auction config/seller experiment group id=].
+
+  Note: Implementations can batch requests by collecting render URLs and ad component render URLs
+  from multiple invocations of [=score and rank a bid=] and passing them all to a single invocation
+  of [=building trusted scoring signals url=] -- the network response has to be parsed to pull out the pieces
+  relevant to each scoreAd() call.
+1. Let |trustedScoringSignals| be null.
+1. Let «|allTrustedScoringSignals|, |scoringDataVersion|» be the result of [=fetching trusted signals=] with
+  |fullSignalsUrl| and false.
+1. If |allTrustedScoringSignals| is an [=ordered map=]:
+  1. Let |trustedScoringSignals| be a new [=map=].
+  1. [=Assert=]: |renderUrls|'s [=set/size=] is 1.
+  1. [=set/For each=] |renderUrl| in |renderUrls|:
+    1. If |allTrustedScoringSignals|["`renderUrls`"] [=map/exists=] and
+      |allTrustedScoringSignals|["`renderUrls`"][|renderUrl|] [=map/exists=]:
+      1. Let |renderUrlValue| be a new [=map=].
+      1. [=map/Set=] |renderUrlValue|[|renderUrl|] to |allTrustedScoringSignals|["`renderUrls`"][|renderUrl|].
+      1. [=map/Set=] |trustedScoringSignals|["`renderUrl`"] to |renderUrlValue|.
+  1. Let |adComponentRenderUrlsValue| be a new [=map=].
+  1. [=set/For each=] |adComponentRenderUrl| in |adComponentRenderUrls|:
+    1. If |allTrustedScoringSignals|["`adComponentRenderUrls`"] [=map/exists=] and
+      |allTrustedScoringSignals|["`adComponentRenderUrls`"][|adComponentRenderUrl|] [=map/exists=]:
+      1. [=map/Set=] |adComponentRenderUrlsValue|[|adComponentRenderUrl|] to
+          |allTrustedScoringSignals|["`adComponentRenderUrls`"][|adComponentRenderUrl|].
+  1. If |adComponentRenderUrlsValue| is not [=map/is empty|empty=]:
+    1. [=map/Set=] |trustedScoringSignals|["`adComponentRenderUrls`"] to |adComponentRenderUrlsValue|.
+1. Let |adMetadata| be |generatedBid|'s [=generated bid/ad=].
 1. Let |bidValue| be |generatedBid|'s [=generated bid/bid=].
 1. If |generatedBid|'s [=generated bid/modified bid=] is not null:
   1. Set |bidValue| to |generatedBid|'s [=generated bid/modified bid=].
-1. TODO: Let |trustedScoringSignals| ... need to assemble and fetch these
 1. Let |browserSignals| be an [=ordered map=] whose [=map/keys=] are [=strings=] and whose
   [=map/values=] are {{any}}.
 1. [=map/Set=] |browserSignals|["`topWindowHostname`"] to [=this=]'s [=relevant settings object=]'s

```

Comment:
This is not part of this PR but we should probably make a TODO/`Issue: Foo` here, because this algorithm is called in parallel so we do not have access to a settings object, and we'll have to get the host from elsewhere.